### PR TITLE
PULL_REQUEST_TEMPLATE: Add encouragement to review PRs

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
As you might have realized, the number of open PR's keeps increasing, we don't have enough people that review, test and merge PRs. This PR hopes to make this a bit better by informing people of this when they open a new PR and encouraging them to review contributions on their own.

If you have rewording suggestions, I'd love to hear them.

This is what I suggested in https://github.com/NixOS/nixpkgs-committers/issues/60#issuecomment-452435126

Ping @edolstra @7c6f434c @alyssais @samueldr @FRidh @Mic92 @domenkozar @Drakonis @aanderse

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (worldofpeace) [inline .github/PULL_REQUEST_TEMPLATE.md]: Do we have a reviewers template for package updates?

Defining a set of things to do and what to look for can make it even clearer for someone to just jump in and start reviewing.

We will also get across what kind of quality we expect from reviewers.

**Comment by worldofpeace** [comment]
Do we have a reviewers template for package updates?

Defining a set of things to do and what to look for can make it even clearer for someone to just jump in and start reviewing.

We will also get across what kind of quality we expect from reviewers.

**Review by 7c6f434c — APPROVED** [review]

**Comment by 7c6f434c** [comment]
I think «test something, report what you tested» might be a good thing to try and see. We have previously seen questions whether all of the checklist items in the PR are mandatory (not sure if we have seen any PR with all of the items from the template checked!). I think once we have enouogh non-committer reviews to want trading quantity for consistency, we can try improving recommendations; I believe this moment is not yet now.

**Comment by worldofpeace** [comment]
> I think «test something, report what you tested» might be a good thing to try and see. We have previously seen questions whether all of the checklist items in the PR are mandatory (not sure if we have seen any PR with all of the items from the template checked!). I think once we have enouogh non-committer reviews to want trading quantity for consistency, we can try improving recommendations; I believe this moment is not yet now.

I think we should encourage at least a baseline of "I've done x things to test and because those were successful this can be merged" and that they've read what changes happened to the software and deduced that there's no breakages.
Without at least that it's very likely, for me personally, to ignore the review because I don't know what they've done to come to a conclusion.

And if they don't have access to the builder pasting the build output showing the success.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (7c6f434c) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
Currently nixpkgs has a lot of new incoming PR's, but not nearly enough people to review this constant stream. Even if you aren't a committer, we would appreciate a review from you, especially for package updates. Just testing the relevant package/service and leaving a comment what and how you tested and whether it worked would be great. See https://github.com/NixOS/nixpkgs/pulls for open PR's. Reviewing PR's isn't mandatory of course, but it would help out the committers immensely and make the average time-to-merge shorter for all of us. Thanks a lot if you do!
```

**Comment by 7c6f434c** [comment]
Argh, GitHub UI is horrible as usual, it doesn't show the actual difference (I added «what you tested»→«what and how you tested»).

For me «what you tested» sounds like «what functionality you tested», but after @worldofpeace comment I understood this can honestly be perceived as «what package you tested» which is indeed of little use.

Pastebinning build output might be a good idea to recommend.

Summarising changelog is something where trusting qualification is indeed needed; maybe we could ask for highlighting any parts of changelog that the commenter finds particularly relevant?

**Comment by worldofpeace** [comment]
> Pastebinning build output might be a good idea to recommend.

I've noticed pr reviewers who do this already in nixpkgs (non commiters) see a quick merge.
They present verifiable proofs to the commiter so that they can personally deduce a change as mergeable. 

 > Summarising changelog is something where trusting qualification is indeed needed; maybe we could ask for highlighting any parts of changelog that the commenter finds particularly relevant?

Yep, the reviewer should read the changelog/release notes and identify items of importance.

I feel like if we could link to https://nixos.org/nixpkgs/manual/#reviewing-contributions-package-updates in the note, and improve that substantially with what we've already coalesced as helpful would give our new reviewers a recipe for success.

**Comment by Infinisil** [comment]
@7c6f434c 

> For me «what you tested» sounds like «what functionality you tested», but after @worldofpeace comment I understood this can honestly be perceived as «what package you tested» which is indeed of little use.

There's lots of PR's that touch multiple packages, such as updates of libraries, in which case it's important to know what was tested.

@worldofpeace 

> Do we have a reviewers template for package updates?
>
> Defining a set of things to do and what to look for can make it even clearer for someone to just jump in and start reviewing.

I'd definitely love for something like this to exist in the future, but until then I think such a simple encouragement in the PR template would be a good start.

> We will also get across what kind of quality we expect from reviewers.

At this point I'd rather have some reviews at all instead of people merging PR's with less and less testing (due to amount of them), causing more breakages in the end. I don't want to scare everybody away by indicating we only want quality reviews.

**Comment by 7c6f434c** [comment]
>There's lots of PR's that touch multiple packages, such as updates of libraries, in which case it's important to know what was tested.
Well, necessary but not sufficient,

**Review by 7c6f434c — APPROVED** [review]

**Comment by worldofpeace** [comment]
> At this point I'd rather have some reviews at all instead of people merging PR's with less and less testing (due to amount of them), causing more breakages in the end. I don't want to scare everybody away by indicating we only want quality reviews.

Substitute quality for formulation. We all already have our formula, that's how we get by.
It just needs to communicated and out in the open.

> I'd definitely love for something like this to exist in the future, but until then I think such a simple encouragement in the PR template would be a good start.

Let's see what we can do, I'm ready for trying to get something out in the works :smile:

**Review by worldofpeace — APPROVED** [review]
No problems with this change. 
It provides immediate exposure to our issue every time someone opens a pr.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (dotlambda) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
Currently nixpkgs has a lot of new incoming PRs, but not nearly enough people to review this constant stream. Even if you aren't a committer, we would appreciate a review from you, especially for package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. See https://github.com/NixOS/nixpkgs/pulls for open PRs. Reviewing PRs isn't mandatory of course, but it would help out the committers immensely and make the average time-to-merge shorter for all of us. Thanks a lot if you do!
```

Comment (dotlambda) [reply .github/PULL_REQUEST_TEMPLATE.md]: ("PR's" -> "PRs")

**Comment by dotlambda** [comment]
How about suggesting to run `nix run nixpkgs.nix-review -c nix-review pr <pr-number>`?

**Comment by veprbl** [comment]
I think we should clarify that we are asking people to contribute a review of *other* peoples PR's.

**Comment by Infinisil** [comment]
@dotlambda Not sure about that. nix-review probably works well for PRs that update a dependency that updates a couple dependent packages. But it won't be practical for PRs that change 100 dependents (nobody wants to build all of those). And it would also encourage users to just run that command and not test the thing properly. Also there's lots of PRs that don't just update packages, for which nix-review can't do anything. So I'd rather keep these instructions very general so they can apply to all PRs and leave room for human decisions for what makes sense to test.

@veprbl I think it's pretty obvious that we don't mean to review their own PRs. And even if not, people testing their own PRs should be done in any case, so there's no harm if they do that (which they'll only consider if they haven't already anyways). Also seeing PR authors tend to their PRs makes it more likely for them to be merged anyways. But if you have a good suggestion for how to word this, I'd love to see it.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (veprbl) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
Currently nixpkgs has a lot of new incoming Pull Requests, but not nearly enough people to review this constant stream. Even if you aren't a committer, we would appreciate a review for a few other PRs from you, especially for package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. See https://github.com/NixOS/nixpkgs/pulls for open PRs. Reviewing PRs isn't mandatory of course, but it would help out the committers immensely and make the average time-to-merge shorter for all of us. Thanks a lot if you do!
```

**Comment by 7c6f434c** [comment]
I would say we could suggest the option of `nix-review` with testing a random subset of rev-deps if either the rebuild amount is low or the person feels like running a long build.

I don't know if we want to remind people to read the diff and test the changes they understand and find probably OK: malicious changes that are dangerous to test are possible, and the risk probably grows with popularity…

**Comment by worldofpeace** [comment]
> I would say we could suggest the option of `nix-review` with testing a random subset of rev-deps if either the rebuild amount is low or the person feels like running a long build.

There's only so much to communicate in an encouragement before you send them on their way, and isn't nix-review in `Things done` now?

**Comment by 7c6f434c** [comment]
>and isn't nix-review in `Things done` now?

Good point, didn't think of that, sorry.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (Infinisil) [reply .github/PULL_REQUEST_TEMPLATE.md]: Pushed a slight variation of this

**Comment by Infinisil** [comment]
I think more detailed descriptions of how and what to test in each scenario could go into the [Reviewing Contributions](https://nixos.org/nixpkgs/manual/#sec-reviewing-contributions) section instead. Maybe we could link that section as a suggestion but make it clear that it's no requirement (to not discourage people, because that section doesn't look very beginner friendly).

**Comment by Infinisil** [comment]
Actually I'm not so sure anymore about linking that section: Right now it only gets updates every 6 months, the current one still mentions `nox-review` for example. I think we're better off not linking to a potentially always outdated manual, especially for that section, because we alter PR workflows more quickly than only exactly every 6 months.

If we had an always up-to-date render of the master/unstable version of the manual I wouldn't mind linking that at all.

**Comment by worldofpeace** [comment]
> Actually I'm not so sure anymore about linking that section: Right now it only gets updates every 6 months, the current one still mentions `nox-review` for example. I think we're better off not linking to a potentially always outdated manual, especially for that section, because we alter PR workflows more quickly than only exactly every 6 months.
> 
> If we had an always up-to-date render of the master/unstable version of the manual I wouldn't mind linking that at all.

I think it makes sense for the nixpkgs manual to always be the version on master.
That's where new contributions go anyways.

Note I've suggested linking to this as well in this thread
> I feel like if we could link to https://nixos.org/nixpkgs/manual/#reviewing-contributions-package-updates in the note, and improve that substantially with what we've already coalesced as helpful would give our new reviewers a recipe for success.

**Comment by 7c6f434c** [comment]
Technically, we have this link: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html 

If we advise people to try the stable version first, then the landing-page manual should be for the stable version, and master version should be linked in contexts intended for contributors.

**Comment by dotlambda** [comment]
> If we advise people to try the stable version first, then the landing-page manual should be for the stable version, and master version should be linked in contexts intended for contributors.

That is certainly true for the NixOS manual but I think it's worth considering a change w.r.t. the nixpkgs one.

**Comment by oxij** [comment]
> Technically, we have this link: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html 

:+1: on using that.

IMHO, both PULL_REQUEST_TEMPLATE and CONTRIBUTING should just link to the manual for the most part and have only GitHub-specific things and checklists in the `.md` files. E.g. I think commit message guidelines and such should live in the nixpkgs manual, not in CONTRIBUTING.

**Comment by Infinisil** [comment]
Sounds good, I added this link. I also encased both links with `<>` as per Apendix E of [RFC 2396](https://www.rfc-editor.org/rfc/rfc2396.txt)

**Comment by Infinisil** [comment]
@dotlambda @oxij Is this PR looking good for you?

**Review by dotlambda — APPROVED** [review]

**Review by 7c6f434c — APPROVED** [review]

**Review by worldofpeace — APPROVED** [review]
Looks good, not sure about the divergence of linking to the unstable manual though.
It's better to link to that one, but we really should have something like that arranged at nixos.org
in the future.

**Review by Ekleog — APPROVED** [review]

**Comment by oxij** [comment]
This would LGTM after a squash, yes.

**Comment by Infinisil** [comment]
Done did the squashing

**Comment by alyssais** [comment]
Does GitHub soft wrap long lines, or will lots of this be cut off?

**Comment by Infinisil** [comment]
I pushed that commit to my master branch, so you can see what it looks like when you make a new PR [here](https://github.com/Infinisil/nixpkgs/compare/master...Infinisil:ports). I think these long spaces after and before the comment should go.

**Comment by Infinisil** [comment]
Removed the big spaces

@alyssais It does wrap them

This is how it currently looks to open a new PR in the default github look: 
unexpanded

I'd like to make this a bit shorter, I'll try rewording it a bit and using link shorteners (any recommendations?)

**Comment by oxij** [comment]
> using link shorteners (any recommendations?)

I would rather it not use any, I vote for adding an unstable manual build to the Makefile of nixos-homepage then: https://github.com/NixOS/nixos-homepage/pull/270.

**Comment by Infinisil** [comment]
I shortened the wording by a lot, such that at least the first heading is visible:
shorter

@oxij Thanks, very cool.

Since I feel like that PR will take some time to be merged though, I think we should merge this one soon in any case.

**Comment by worldofpeace** [comment]
Can we do the
```
<!---
```
and 
```
--->
```

on newlines?

**Comment by Infinisil** [comment]
@worldofpeace I removed this in recent commits, because it only takes up space, and I'd like for users to at least see the "Motivation for this change" line by default. With these on separate lines that's almost impossible.

**Comment by Infinisil** [comment]
@worldofpeace Is it alright like this or do you think we need those newlines?

**Comment by worldofpeace** [comment]
> @worldofpeace Is it alright like this or do you think we need those newlines?

This change has looked good for a while. How much can newlines help with right :smile:
##### Though it would make it look less flowy' because its grouped between something.

So I think we should be good to merge :+1:

**Review by worldofpeace — APPROVED** [review]

**Comment by Ekleog** [comment]
Seeing @timokau's discourse thread linked above, maybe it would make sense to add a link to the discourse thread in the PR template? Not sure what @timokau thinks about it yet, though, as it may be better to start that discourse thread experiment without too much incentive.

Now, I'm wondering whether discourse is a proper UI, as if multiple committers monitor the thread there is no good way of marking a comment over there as “done” so only one needs to click the link I think, but…

**Comment by timokau** [comment]
> Seeing @timokau's discourse thread linked above, maybe it would make sense to add a link to the discourse thread in the PR template? Not sure what @timokau thinks about it yet, though, as it may be better to start that discourse thread experiment without too much incentive.

I'd be happy with that, but maybe give it a few ways to see how it goes :)

> Now, I'm wondering whether discourse is a proper UI, as if multiple committers monitor the thread there is no good way of marking a comment over there as “done” so only one needs to click the link I think, but…

https://discourse.nixos.org/t/discussion-pr-reviews/2619

**Pull request merged**
